service/history/workflow/retry.go

```diff
@@ -35,22 +35,37 @@ import (
 	"go.temporal.io/server/common/primitives/timestamp"
 )
 
+// TODO treat 0 as 0, not infinite
+
 func getBackoffInterval(
 	now time.Time,
-	expirationTime time.Time,
-	currentAttemptCounterValue int32,
+	currentAttempt int32,
 	maxAttempts int32,
 	initInterval *time.Duration,
 	maxInterval *time.Duration,
+	expirationTime *time.Time,
 	backoffCoefficient float64,
 	failure *failurepb.Failure,
 	nonRetryableTypes []string,
 ) (time.Duration, enumspb.RetryState) {
 
-	// Sanity check to make sure currentAttemptCounterValue started with 1.
-	if currentAttemptCounterValue < 1 {
-		currentAttemptCounterValue = 1
+	// TODO remove below checks, most are already set with correct values
+	if currentAttempt < 1 {
+		currentAttempt = 1
+	}
+
+	if initInterval == nil {
+		initInterval = timestamp.DurationPtr(time.Duration(0))
```

Comment:
initial interval set to 0 would cause trouble because backoff is based on initial interval. I believe the default we set is 1s

Reply:
there is already a default set when creating the activity at the beginning.
https://github.com/temporalio/temporal/blob/v1.10.4/service/history/commandChecker.go#L671
https://github.com/temporalio/temporal/blob/v1.10.4/common/util.go#L485
https://github.com/temporalio/temporal/blob/v1.10.4/common/util.go#L89

Reply:
Yes, that is my point. The default is 1s. The nil case should not happen, so we either throw when we see the nil, or we set it to default of 1s. Set to 0 will cause other trouble and hide the issue. For example, based on the logic below, 0 interval might cause no retry.

Reply:
overhaul of code is required to making things right
* `0 is 0, nil is nil is required`
* rely on input validation to correctly set attributes and later use the variables without overriding (sanity check & panic should be allowed)

this PR is going to be cherry picked for a patch release, so prefer minimal changes.

Reply:
So both `0` and `nil` are not valid here. I would suggest:
```suggestion
       if initInterval == nil || *initialInterval == 0 {
		return backoff.NoBackoff, enumspb.RETRY_STATE_RETRY_POLICY_NOT_SET
```

Reply:
the code below will either override init interval to max interval (if not nil) or just return no retry